synapse/storage/monthly_active_users.py

```diff
@@ -172,26 +174,36 @@ def get_registered_reserved_users_count(self):
 
     @defer.inlineCallbacks
     def upsert_monthly_active_user(self, user_id):
-        """Updates or inserts monthly active user member
-        Arguments:
+        """Updates or inserts the user into the monthly active user table, which
+        is used to track the current MAU usage of the server
+
+        Args:
             user_id (str): user to add/update
         """
         is_insert = yield self.runInteraction(
             "upsert_monthly_active_user", self.upsert_monthly_active_user_txn,
             user_id
         )
+        # Considered pushing cache invalidation down into txn method, but
```

Comment:
sorry, this isn't terribly easy to read; and I think it's actually more important to put something in the `_txn` method. How about getting rid of this, and instead, putting the following in the docstring of the `_txn` method:

> Note that, after calling this method, it will generally be necessary to invalidate the caches on `user_last_seen_monthly_active` and `get_monthly_active_count`. We can't do that here, because we are running in a database thread rather than the main thread, and we can't call `txn.call_after` because `txn` may not be a `LoggingTransaction`.

Reply:
wfm